compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirEqualityOperatorCallChecker.kt

```diff
@@ -0,0 +1,77 @@
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.expressions.FirEqualityOperatorCall
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.resolve.inference.inferenceComponents
+import org.jetbrains.kotlin.fir.resolve.toFirRegularClass
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.fir.types.ConeTypeCompatibilityChecker.areCompatible
+
+object FirEqualityOperatorCallChecker : FirBasicExpressionChecker() {
+    override fun check(expression: FirStatement, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirEqualityOperatorCall) return
+        val arguments = expression.argumentList.arguments
+        if (arguments.size != 2) return
+        val lValue = arguments[0]
+        val rValue = arguments[1]
+        val lType = lValue.typeRef.coneType
+        val rType = rValue.typeRef.coneType
+        // If one of the type is already nothing, we skip reporting further comparison. This is to allow comparing with `null`, which has
+        // type `Nothing?`
+        if (lType.isNullableNothing || rType.isNullableNothing) return
+        val inferenceContext = context.session.inferenceComponents.ctx
+        val intersectionType = inferenceContext.intersectTypesOrNull(listOf(lType, rType)) as? ConeIntersectionType ?: return
+
+        if (!intersectionType.intersectedTypes.areCompatible(inferenceContext)) {
+            // Note: FE1.0 reports INCOMPATIBLE_ENUM_COMPARISON_ERROR only when TypeIntersector.isIntersectionEmpty() thinks the given types
+            // are compatible. Exactly mimicking the behavior of FE1.0 is difficult and does not seem to provide any value. So instead,
+            // we deterministically output INCOMPATIBLE_ENUM_COMPARISON_ERROR if the code is comparing two enums.
+            if (lType.isEnumType(context) && rType.isEnumType(context)) {
+                reporter.reportOn(
+                    expression.source,
+                    FirErrors.INCOMPATIBLE_ENUM_COMPARISON_ERROR,
+                    lType,
+                    rType,
+                    context
+                )
+            } else if (expression.source?.kind == FirRealSourceElementKind) {
+                reporter.reportOn(
+                    expression.source,
+                    FirErrors.EQUALITY_NOT_APPLICABLE,
+                    expression.operation.operator,
+                    lType,
+                    rType,
+                    context
+                )
+            } else if (expression.source?.kind == FirFakeSourceElementKind.WhenCondition) {
+                reporter.reportOn(
+                    expression.source,
+                    FirErrors.INCOMPATIBLE_TYPES,
```

Comment:
Is this checker meant to cover all `INCOMPATIBLE_TYPES` errors? I see other tests with `INCOMPATIBLE_TYPES` (E.g., `diagnostics/tests/IsExpressions.kt`) where it's not an equality operator.

Reply:
No it's not meant to cover all. It's only supposed to cover `FirEqualityOperatorCall`.